tests/unit/TestNetRemoteDataStreamingServiceClient.cxx

```diff
@@ -0,0 +1,68 @@
+
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+#include <catch2/catch_test_macros.hpp>
+#include <grpcpp/create_channel.h>
+#include <microsoft/net/remote/NetRemoteServer.hxx>
+#include <microsoft/net/remote/protocol/NetRemoteDataStreamingService.grpc.pb.h>
+
+#include "TestNetRemoteCommon.hxx"
+#include "TestNetRemoteDataStreamingReactors.hxx"
+
+TEST_CASE("DataStreamUpload API", "[basic][rpc][client][remote][stream]")
+{
+    using namespace Microsoft::Net::Remote;
+    using namespace Microsoft::Net::Remote::DataStream;
+    using namespace Microsoft::Net::Remote::Service;
+
+    using Microsoft::Net::Remote::Test::DataStreamWriter;
+    using Microsoft::Net::Remote::Test::RemoteServiceAddressHttp;
+
+    NetRemoteServerConfiguration Configuration{
+        .ServerAddress = RemoteServiceAddressHttp,
+    };
+
+    NetRemoteServer server{ Configuration };
+    server.Run();
+
+    auto channel = grpc::CreateChannel(RemoteServiceAddressHttp, grpc::InsecureChannelCredentials());
+    auto client = NetRemoteDataStreaming::NewStub(channel);
+
+    static constexpr auto numberOfDataBlocksToWrite = 10;
+
+    SECTION("Can be called with one client")
+    {
+        auto dataStreamWriter = std::make_unique<DataStreamWriter>(client.get(), numberOfDataBlocksToWrite);
+
+        DataStreamUploadResult result{};
+        grpc::Status status = dataStreamWriter->Await(&result);
+        REQUIRE(status.ok());
+        REQUIRE(result.numberofdatablocksreceived() == numberOfDataBlocksToWrite);
+        REQUIRE(result.status().code() == DataStreamOperationStatusCodeSucceeded);
+    }
+
+    SECTION("Can be called with multiple parallel clients")
+    {
+        static constexpr auto numberOfClients = 5;
+
+        std::vector<std::unique_ptr<NetRemoteDataStreaming::Stub>> dataStreamingClients;
+        std::vector<std::jthread> clientThreads;
+
+        for (std::size_t i = 0; i < numberOfClients; i++) {
+            dataStreamingClients.push_back(NetRemoteDataStreaming::NewStub(channel));
+            auto dataStreamWriter = std::make_unique<DataStreamWriter>(dataStreamingClients.back().get(), numberOfDataBlocksToWrite);
+
+            clientThreads.emplace_back([dataStreamWriter = std::move(dataStreamWriter)]() {
+                DataStreamUploadResult result{};
+                grpc::Status status = dataStreamWriter->Await(&result);
+                REQUIRE(status.ok());
+                REQUIRE(result.numberofdatablocksreceived() == numberOfDataBlocksToWrite);
+                REQUIRE(result.status().code() == DataStreamOperationStatusCodeSucceeded);
+            });
+        }
+    }
```

Comment:
I think we should verify that the streaming threads don't throw any exceptions. Right now, when the threads are destroyed at end of scope, their destructors will call `join()` automatically, but if an exception is thrown, `std::terminate` will be called and the test will crash. Instead, we should call `join` explicitly and verify no exceptions are thrown.

```suggestion
        }
        for (auto& clientThread : clientThreads) {
            REQUIRE_NOTHROW(clientThread.join());
        }
    }
```